### PR TITLE
create a readonly database role

### PR DESCRIPTION
**Pull request by kyscott18**

### CODE REVIEW

File: packages/core/src/database/index.ts [hunk 1]
```diff
@@ -215,22 +232,64 @@ export const createDatabase = ({
...
+    await internal.query(`CREATE SCHEMA IF NOT EXISTS "${preBuild.namespace}"`);
+    const hasRole = await internal
+      .query("SELECT FROM pg_roles WHERE rolname = $1", [role])
+      .then(({ rows }) => rows[0]);
+    if (hasRole) {
+      await internal.query(`DROP OWNED BY ${role}`);
+      await internal.query(`DROP ROLE IF EXISTS ${role}`);
+    }
+    await internal.query(
+      `CREATE ROLE ${role} WITH LOGIN PASSWORD '${connection.password}'`,
```
Comment (typedarray) [inline packages/core/src/database/index.ts:257]: We should probably use a different password than the one from the main/admin/superuser connection string. Looks like PG requires a non-null/non-empty password, so we could use something like "public" or "readonly", or generate a less predictable one. cc @jaylmiller

File: packages/core/src/database/index.ts [hunk 2]
```diff
@@ -215,22 +232,64 @@ export const createDatabase = ({
...
+    );
+    await internal.query(
+      `GRANT CONNECT ON DATABASE "${connection.database}" TO ${role}`,
+    );
+    await internal.query(`REVOKE ALL ON SCHEMA public FROM ${role}`);
```
Comment (typedarray) [inline packages/core/src/database/index.ts:262]: Will this fail if the `public` schema doesn't exist?

File: packages/core/src/database/index.ts [hunk 3]
```diff
@@ -106,30 +106,34 @@ type PGliteDriver = {
...
-  /** Used in api functions */
-  readonly: HeadlessKysely<unknown>;
   /** Used to interact with the sync-store */
   sync: HeadlessKysely<PonderSyncSchema>;
+  /** Used in api functions */
+  readonly: HeadlessKysely<unknown>;
+  /** Used in client queries */
+  // client: HeadlessKysely<unknown>;
+  drizzle: Drizzle<Schema>;
+  drizzleClient: Drizzle<Schema>;
```
Comment (typedarray) [inline packages/core/src/database/index.ts:125]: Would find this clearer as `drizzleReadonly`, "client" makes it seem like it's a database client/connection object (vs a pool).

**Review by typedarray — COMMENTED** [review]
Looking great

File: packages/core/src/database/index.ts (hunk 1 above)
Comment (jaylmiller) [reply packages/core/src/database/index.ts:257]: default to something predictable but make option to override would be good

File: packages/core/src/database/index.ts [hunk 4]
```diff
@@ -215,22 +232,64 @@ export const createDatabase = ({
...
+    );
+    await internal.query(
+      `GRANT CONNECT ON DATABASE "${connection.database}" TO ${role}`,
+    );
+    await internal.query(`REVOKE ALL ON SCHEMA public FROM ${role}`);
+    await internal.query(
+      `REVOKE ALL ON ALL TABLES IN SCHEMA public FROM ${role}`,
+    );
+    await internal.query(
+      `GRANT USAGE ON SCHEMA ${preBuild.namespace} TO ${role}`,
```
Comment (jaylmiller) [inline packages/core/src/database/index.ts:267]: add `"` around namespace

File: packages/core/src/database/index.ts [hunk 5]
```diff
@@ -204,6 +213,14 @@ export const createDatabase = ({
         },
         plugins: [new WithSchemaPlugin("ponder_sync")],
       }),
+      drizzle: drizzlePglite((driver as PGliteDriver).instance, {
+        casing: "snake_case",
```
Comment (typedarray) [inline packages/core/src/database/index.ts:217]: Is this not the default?

File: packages/core/src/database/index.ts [hunk 6]
```diff
@@ -151,6 +155,11 @@ export const createDatabase = ({
 
     const kyselyDialect = new KyselyPGlite(driver.instance).dialect;
 
+    await driver.instance.query(
+      `CREATE SCHEMA IF NOT EXISTS "${preBuild.namespace}"`,
+    );
+    await driver.instance.query(`SET search_path TO ${preBuild.namespace}`);
```
Comment (typedarray) [inline packages/core/src/database/index.ts:161]: Do we need to escape the namespace here with double quotes? Same question for a few queries below in the role setup.

File: packages/core/src/database/index.ts [hunk 7]
```diff
@@ -215,22 +232,58 @@ export const createDatabase = ({
...
+    await internal.query(
+      `GRANT CONNECT ON DATABASE "${connection.database}" TO ${role}`,
+    );
+    await internal.query(
+      `GRANT USAGE ON SCHEMA "${preBuild.namespace}" TO ${role}`,
+    );
+    await internal.query(
+      `ALTER DEFAULT PRIVILEGES IN SCHEMA "${preBuild.namespace}" GRANT SELECT ON TABLES TO ${role}`,
+    );
+    await internal.query(
```
Comment (typedarray) [inline packages/core/src/database/index.ts:266]: Is there a way to do a multi-statement raw query here? Would be easier to read and might be faster.

File: packages/core/src/database/index.ts [hunk 8]
```diff
@@ -215,22 +232,58 @@ export const createDatabase = ({
...
       readonly: createPool({
         ...preBuild.databaseConfig.poolConfig,
+        connectionString: undefined,
         application_name: `${preBuild.namespace}_readonly`,
         max: readonlyMax,
+        user: role,
+        password: "pw",
+        host: connection.host ?? undefined,
+        port: Number(connection.port!),
+        database: connection.database ?? undefined,
```
Comment (typedarray) [inline packages/core/src/database/index.ts:286]: If this is actually nullable, we might have a problem above in the role name. But I'm pretty sure PG conn strings must explicitly set a database name, so might just be a bad type.

File: packages/core/src/database/index.ts [hunk 9]
```diff
@@ -106,30 +106,34 @@ type PGliteDriver = {
...
+  /** Used in api functions */
+  readonly: HeadlessKysely<unknown>;
+  /** Used in client queries */
+  // client: HeadlessKysely<unknown>;
+  drizzle: Drizzle<Schema>;
+  drizzleReadonly: Drizzle<Schema>;
 };
 
-export const createDatabase = ({
+export const createDatabase = async ({
```
Comment (typedarray) [inline packages/core/src/database/index.ts:128]: Small nit - it's getting difficult to remember what happens in `createDatabase` vs `database.setup`. Maybe need a new name for setup - `prepareNamespace` ?

File: packages/core/src/database/index.ts (hunk 7 above)
Comment (kyscott18) [reply packages/core/src/database/index.ts:266]: I don't think so, from what I understand the client only allows one statement per query

File: packages/core/src/database/index.ts (hunk 5 above)
Comment (kyscott18) [reply packages/core/src/database/index.ts:217]: I don't think it is

**Pull request merged**